src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs

```diff
@@ -264,76 +261,61 @@ public static LinePositionSpan GetLinePositionSpan(DiagnosticDataLocation? dataL
-            SwapIfNeeded(ref startLinePosition, ref endLinePosition);
 
-            return new LinePositionSpan(startLinePosition, endLinePosition);
-        }
-
-        public static TextSpan GetTextSpan(DiagnosticDataLocation? dataLocation, SourceText text)
-        {
-            var linePositionSpan = GetLinePositionSpan(dataLocation, text, useMapped: false);
-
-            var span = text.Lines.GetTextSpan(linePositionSpan);
-            return EnsureInBounds(TextSpan.FromBounds(Math.Max(span.Start, 0), Math.Max(span.End, 0)), text);
-        }
+            var startLine = (useMapped ? dataLocation?.MappedStartLine : dataLocation?.OriginalStartLine) ?? 0;
+            var endLine = (useMapped ? dataLocation?.MappedEndLine : dataLocation?.OriginalEndLine) ?? 0;
 
-        private static void AdjustBoundaries(int dataLocationStartLine, int dataLocationStartColumn, int dataLocationEndLine, int dataLocationEndColumn,
-            TextLineCollection lines, out int startLine, out int startColumn, out int endLine, out int endColumn)
-        {
-            startLine = dataLocationStartLine;
-            var originalStartColumn = dataLocationStartColumn;
+            // Make sure the starting columns are never negative.
+            var startColumn = Math.Max((useMapped ? dataLocation?.MappedStartColumn : dataLocation?.OriginalStartColumn) ?? 0, 0);
+            var endColumn = Math.Max((useMapped ? dataLocation?.MappedEndColumn : dataLocation?.OriginalEndColumn) ?? 0, 0);
 
-            startColumn = Math.Max(originalStartColumn, 0);
             if (startLine < 0)
             {
+                // If the start line is negative (e.g. before the start of the actual document) then move the start to the 0,0 position.
                 startLine = 0;
                 startColumn = 0;
             }
+            else if (startLine >= lines.Count)
+            {
+                // if the start line is after the end of the document, move the start to the last location in the document.
+                startLine = lines.Count - 1;
+                startColumn = lines[startLine].SpanIncludingLineBreak.Length;
+            }
 
-            endLine = dataLocationEndLine;
-            var originalEndColumn = dataLocationEndColumn;
-
-            endColumn = Math.Max(originalEndColumn, 0);
             if (endLine < 0)
             {
+                // if the end is before the start of the document, then move the end to wherever the start position was determined to be.
                 endLine = startLine;
                 endColumn = startColumn;
             }
             else if (endLine >= lines.Count)
             {
+                // if the end line is after the end of the document, move the end to the last location in the document.
                 endLine = lines.Count - 1;
-                endColumn = lines[endLine].EndIncludingLineBreak;
+                endColumn = lines[endLine].SpanIncludingLineBreak.Length;
```

Comment:
prior logic was completely broken.  endColumn is from 0 to length-of-line.   EndIncludingLineBreak is the absolute position of hte end of hte line in the file.